aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+    logger.info(f'Inserting {nrows}/{len(obj)} rows')
+
+    # Insert the new rows
+    job_config = bq.LoadJobConfig()
+    job_config.source_format = bq.SourceFormat.NEWLINE_DELIMITED_JSON
+    job_config.schema = get_formatted_bq_schema()
+
+    j = '\n'.join(json.dumps(o) for o in filtered_obj)
+
+    resp = bigquery_client.load_table_from_file(
+        StringIO(j), table, job_config=job_config
+    )
+    try:
+        result = resp.result()
+        logger.info(f'Inserted {result.output_rows}/{nrows} rows')
+    except ClientError as e:
+        logger.error(resp.errors)
+        raise e
+
+    return nrows
+
+
+def insert_dataframe_rows_in_table(table: str, df: pd.DataFrame):
+    """Insert new rows from dataframe into a table"""
+
+    _query = f"""
+        SELECT id FROM `{table}`
+        WHERE id IN UNNEST(@ids);
+    """
+    job_config = bq.QueryJobConfig(
+        query_parameters=[
+            bq.ArrayQueryParameter('ids', 'STRING', list(set(df['id']))),
+        ]
+    )
+    result = bigquery_client.query(_query, job_config=job_config).result()
+    existing_ids = set(result.to_dataframe()['id'])
+
+    # Filter out any rows that are already in the table
+    df = df[~df['id'].isin(existing_ids)]
+
+    # Count number of rows adding
+    adding_rows = len(df)
+
+    # Insert the new rows
+    project_id = table.split('.')[0]
+    table_schema = get_bq_schema_json()
+
+    df.to_gbq(
+        table,
+        project_id=project_id,
+        table_schema=table_schema,
+        if_exists='append',
+    )
+
+    logger.info(f'{adding_rows} new rows inserted')
+    return adding_rows
+
+
+CACHED_CURRENCY_CONVERSION: Dict[str, float] = {}
+
+
+def get_currency_conversion_rate_for_time(time: datetime):
+    """
+    Get the currency conversion rate for a given time.
+    Noting that GCP conversion rates are decided at the start of the month,
+    and apply to each job that starts within the month, regardless of when
+    the job finishes.
+    """
+
+    key = f'{time.year}-{time.month}'
+    if key not in CACHED_CURRENCY_CONVERSION:
+        logger.info(f'Looking up currency conversion rate for {key}')
+        query = f"""
+            SELECT currency_conversion_rate
+            FROM {GCP_BILLING_BQ_TABLE}
+            WHERE DATE(_PARTITIONTIME) = DATE('{time.date()}')
+            LIMIT 1
+        """
+        for r in bigquery_client.query(query).result():
+            CACHED_CURRENCY_CONVERSION[key] = r['currency_conversion_rate']
+
+    return CACHED_CURRENCY_CONVERSION[key]
+
+
+def _generate_hail_resource_cost_lookup():
+    """
+    Generate the cost table for the hail resources.
+    This is currently set to the australia-southeast-1 region.
+    """
+    # pylint: disable=import-outside-toplevel,import-error
+    from hailtop.utils import (
+        rate_gib_hour_to_mib_msec,
+        rate_gib_month_to_mib_msec,
+        rate_cpu_hour_to_mcpu_msec,
+        rate_instance_hour_to_fraction_msec,
+    )
+
+    rates = [
```

Comment:
We could consider doing this in the future: https://github.com/hail-is/hail/pull/11840/files

Reply:
Noting that this function doesn't handle rate changes over time, they're just static